File: runtime/erlang/src/runtime_http_client.erl
%% HTTP client behaviour and shared request-building helpers.
-module(runtime_http_client).
-include("runtime_types.hrl").
-include("runtime_http_client.hrl").

-export([
    build_url/2,
    build_request/2,
    method_atom/1,
    content_type/1
]).

-callback request(http_client_request()) -> {ok, http_response()} | {error, term()}.

-spec build_url(#{binary() => term()}, http_request()) -> binary().
build_url(Config, #http_request{path = Path, query = Query, host = Host}) ->
    BaseUrl = maps:get(base_url, Config, undefined),
    QueryStr =
        case maps:to_list(Query) of
            [] ->
                <<>>;
            Pairs ->
                Encoded = uri_string:compose_query([{K, V} || {K, V} <- Pairs]),
                <<"?", Encoded/binary>>
        end,
    {Scheme, DefaultAuthority} = runtime_utils:split_base_url(BaseUrl),
    Authority =
        case Host of
            undefined -> DefaultAuthority;
            _ -> Host
        end,
    <<Scheme/binary, Authority/binary, Path/binary, QueryStr/binary>>.

-spec build_request(#{binary() => term()}, http_request()) -> http_client_request().
build_request(Config, #http_request{} = Request) ->
    #http_client_request{
        method = method_atom(Request#http_request.method),
        url = build_url(Config, Request),
        headers = Request#http_request.headers,
        body = Request#http_request.body
    }.

-spec method_atom(binary()) -> atom().
method_atom(Method) when is_binary(Method) ->
    binary_to_atom(string:lowercase(Method), utf8).

-spec content_type([{binary(), binary()}]) -> binary().
content_type(Headers) ->
    case proplists:get_value(<<"Content-Type">>, Headers) of
        undefined -> <<"application/octet-stream">>;
        CT -> CT
    end.

File: runtime/elixir/lib/runtime_http_client.ex
defmodule RuntimeHttpClient.Request do
  @moduledoc false

  @type t :: %__MODULE__{
          method: atom(),
          url: String.t(),
          headers: [{String.t(), String.t()}],
          body: iodata()
        }

  defstruct method: nil,
            url: "",
            headers: [],
            body: ""
end

defmodule RuntimeHttpClient do
  @moduledoc "HTTP client behaviour and shared request-building helpers."

  alias RuntimeHttpClient.Request
  alias RuntimeTypes.HttpRequest
  alias RuntimeTypes.HttpResponse
  alias RuntimeUtils

  @callback request(Request.t()) :: {:ok, HttpResponse.t()} | {:error, term()}

  @spec build_url(map(), HttpRequest.t()) :: String.t()
  def build_url(config, %HttpRequest{path: path, query: query, host: host}) do
    base_url = Map.get(config, :base_url)

    query_str =
      case Map.to_list(query) do
        [] -> ""
        pairs -> "?" <> URI.encode_query(pairs)
      end

    {scheme, default_authority} = RuntimeUtils.split_base_url(base_url || "")

    authority =
      case host do
        nil -> default_authority
        value -> value
      end

    scheme <> authority <> path <> query_str
  end

  @spec build_request(map(), HttpRequest.t()) :: Request.t()
  def build_request(config, %HttpRequest{} = req) do
    %Request{
      method: method_atom(req.method),
      url: build_url(config, req),
      headers: req.headers,
      body: req.body
    }
  end

  @spec method_atom(String.t()) :: atom()
  def method_atom(method) when is_binary(method) do
    method |> String.downcase() |> String.to_atom()
  end

  @spec content_type([{String.t(), String.t()}]) :: String.t()
  def content_type(headers) do
    case List.keyfind(headers, "Content-Type", 0) do
      {"Content-Type", value} -> value
      nil -> "application/octet-stream"
    end
  end
end
